### PR TITLE
[llvm-readobj] Fixes malformed json on JSON printed corefiles

### PR DESCRIPTION
**Pull request by feg208**
This patch fixes an issue where, when printing corefile notes with llvm-readobj as json, the dumper generated llvm formatted output which isn't valid json. This alters the dumper to, in the NT_FILE, note, dump properly formatted json data.

Prior to this patch the JSON output was formatted like:
 ```
                "Mapping": [
                  "Start": 4096,
                  "End": 8192,
                  "Offset": 12288,
                  "Filename": "/path/to/a.out"
                ],
 ```
 Whereas now it is formatted as:
  ```
                "Mappings": [
                  {
                    "Start": 4096,
                    "End": 8192,
                    "Offset": 12288,
                    "Filename": "/path/to/a.out"
                  },
  ```
  Which is valid. Additionally the LLVM output has changed to match the structure of the JSON output (i.e. instead of lists of keys it is a list of dictionaries)

### CODE REVIEW

File: llvm/test/tools/llvm-readobj/ELF/note-core-ntfile.test
Comment (jh7370) [inline llvm/test/tools/llvm-readobj/ELF/note-core-ntfile.test:7]: Could you check prior art in other tests, please, but I don't think there's any value in `llvm-readelf` and  `llvm-readobj` invocations where both use `--elf-output-style=JSON`.

**Review by jh7370 — COMMENTED** [review]
Rather than adding a near-duplicate method to work around the fact that the existing code doesn't do what we need it to for JSON output, could we instead simply change the existing code to produce output that is valid (i.e. replace the existing function with the new one you've written)? It would mean that the LLVM output would change, but I don't think this is an issue: as far as I'm aware, we don't promise full output compatibility between LLVM releases of the LLVM-style output. We'd want a release note in this case, to advise of both the JSON fix and the change in LLVM style.

**Comment by feg208** [comment]
>  It would mean that the LLVM output would change, but I don't think this is an issue: as far as I'm aware, we don't promise full output compatibility between LLVM releases of the LLVM-style output. We'd want a release note in this case, to advise of both the JSON fix and the change in LLVM style.

Sure. As long as that is ok. I didn't really know what the expectation for the llvm format was.

File: llvm/test/tools/llvm-readobj/ELF/note-core-ntfile.test
Comment (feg208) [reply llvm/test/tools/llvm-readobj/ELF/note-core-ntfile.test:7]: seems like the practice is to omit readelf

**Comment by feg208** [comment]
@jh7370 I made the recommended changes and updated the llvm style test

**Review by ilovepi — APPROVED** [review]
This looks good from my perspective. and I agree w/ @jh7370 about the compatibility and approach. Please do wait for his LGTM before landing, though.

Can you also include an example of the bad vs. corrected JSON in the commit message? Something minimal to show what's changed is probably fine.

**Comment by feg208** [comment]
> Can you also include an example of the bad vs. corrected JSON in the commit message? Something minimal to show what's changed is probably fine.

Done!

**Comment by ilovepi** [comment]
Totally up to you, but I'd just show the relevant bits.

```JSON
"Mapping": [
     "Start": 4096,
       "End": 8192,
        "Offset": 12288,
         "Filename": "/path/to/a.out"
],
```

```JSON
"Mappings": [
  {
     "Start": 4096,
      "End": 8192,
       "Offset": 12288,
          "Filename": "/path/to/a.out"
  },
 ...
]
```

**Review by jh7370 — COMMENTED** [review]
Code looks good, barring a doc issue. Please also update the PR description to match the current change, in particular noting that the LLVM output has been structurally changed to make it work for JSON output too.

**Comment by feg208** [comment]
> Code looks good, barring a doc issue. Please also update the PR description to match the current change, in particular noting that the LLVM output has been structurally changed to make it work for JSON output too.

@jh7370 I altered the documentation and shortened the commit message along the lines laid out in @ilovepi's comments

File: llvm/docs/ReleaseNotes.rst
Comment (jh7370) [inline llvm/docs/ReleaseNotes.rst:241]: Sorry, should have noticed this earlier - you might as well mention the bug fix at the same time. Also, if you look at the grammar of other release notes, you'll notice this note isn't quite consistent.
```suggestion
* llvm-readobj's LLVM output format for ELF core files has been changed. Similarly, the JSON format has been fixed for this case. The
```
(make sure to reflow as necessary at the line limit)

**Review by jh7370 — APPROVED** [review]
In the PR description:
> NT_FILE, note use

I think "note" here should be "now".

LGTM, with that and the other inline comment.

File: llvm/docs/ReleaseNotes.rst
Comment (feg208) [reply llvm/docs/ReleaseNotes.rst]: done

**Pull request merged**
